FUSETOOLS-2873 - adding model elements and property pages for REST

bfitzpat opened the pull request:
Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

# Pull Request Checklist

After this checklist is all checked or PR provides explanations for possible pass-through, please put the label "Ready for review" on the PR.


## General

- [ ] Did you use the Jira Issue number in the commit comments?
- [ ] Did you set meaningful commit comments on each commit?
- [ ] Did you sign off all commits for this PR? (git commit -s -m "jira issue number - your commit comment")

## Functional

- [ ] Did the CI job report a successful build?
- [ ] Is the non-happy path working, too?

## Maintainability

- [ ] Are all Sonar reported issues fixed or can they be ignored?
- [ ] Is there no duplicated code?
- [ ] Are method-/class-/variable-names meaningful?

## Tests

- [ ] Are there unit-tests?
- [ ] Are there integration tests (or at least a jira to tackle these)?
- [ ] Do we need a new UI test?

## Legal

- [ ] Have you used the correct file header copyright comment?
- [ ] Have you used the correct year in the headers of new files?

bfitzpat commented:
This PR now does the following:
* generates the model upstream in the CamelContextElement model object with help from the new RestConfigurationElement, RestElement, and RestVerbElement objects
* populates the REST tab in the editor from the main model
* adds a property page for the selected REST element, which is currently not read only
* adds a property page for the selected REST operation, which is currently not read only

Issues include:
* need to add IDs for all REST elements and REST operations without them, otherwise not all objects (REST elements or operations) appear in the correct lists

Looking for feedback on the approach I used as well as how to fix the ID issue correctly, when @lhein or @apupier get back to their desks next week.

bfitzpat commented:
Added first cut of support for "To" property (corresponds to child "To" element for REST operations). Doesn't quite work, so curious how to get it working. :)

apupier commented on core/plugins/org.fusesource.ide.camel.model.service.core/src/org/fusesource/ide/camel/model/service/core/model/CamelContextElement.java:
```diff
@@ -38,6 +38,16 @@
+	/**
+	 * Contains REST configuration elements
+	 */
+	private Map<String, AbstractCamelModelElement> restConfigurations = new HashMap<>();
```
The mentioned problem with the mandatory ids is coming from the implementation of the model which was based on the same pattern than for global elements, which are requiring ids.

bfitzpat replied:
We are now generating IDs for the REST element and all REST operations (verbs)

apupier commented on core/plugins/org.fusesource.ide.camel.model.service.core/src/org/fusesource/ide/camel/model/service/core/model/RestConfigurationElement.java:
```diff
@@ -0,0 +1,88 @@
+	 */
+	public RestConfigurationElement(AbstractCamelModelElement parent, Node underlyingNode) {
+		super(parent, underlyingNode);
+//		setUnderlyingMetaModelObject(new GlobalBeanEIP());
```
commented out code to remove

apupier commented:
> Added first cut of support for "To" property (corresponds to child "To" element for REST operations). Doesn't quite work, so curious how to get it working. :)

Can you clarify what you are talking about?
Also isn't a quite separated feature than the other ones? Does it make sense to create a separated PR?

apupier commented:
I created https://github.com/bfitzpat/fuseide/pull/18 for investigating non-mandatory ids.
I created https://github.com/bfitzpat/fuseide/pull/17 for proposing improvements

bfitzpat commented:
@apupier and @lhein 

I think I have addressed most of the issues that have come up with this. We now have a read only way of looking at the REST Configuration, REST Element, and REST Operations/Verbs. I have corrected the reported issues from Sonar and cleaned up some of the commented out code no longer needed. 

The one question I have left is regarding model updates.

I anticipate that users will go to the source tab to edit their REST DSL XML. When they do so, that should update the underlying model. In the RestConfigurationEditor, line 158, I'm adding a model listener but it never gets triggered so far as I can tell. Even going back to CamelFile, line 429 (fireModelChanged) that doesn't seem to be fired when the source tab changes in debug mode.

What's the recommended pattern here for listening for changes on other pages of the editor and refreshing the Rest page?

bfitzpat commented:
That is a bizarre test failure.

apupier commented:
> That is a bizarre test failure.
> 
> Same tests run just fine locally at command line with maven. (i.e. build root (mvn clean verify -DskipTests, change to fuseide\core\tests\org.fusesource.ide.camel.validation.tests.integration directory and run mvn install.

from latest link that I found https://fusesource-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/jbosstools-fuse_PullRequest/1150/testReport/junit/org.fusesource.ide.camel.validator.marker/MarkersRemoverValidatorIT/checkMarkers/

the failing test is:
java.lang.AssertionError: 
Expected size:<2> but was:<0> in:
<[]>
	at org.fusesource.ide.camel.validator.marker.MarkersRemoverValidatorIT.checkMarkers(MarkersRemoverValidatorIT.java:81)

it is is a known unstable test https://issues.jboss.org/browse/FUSETOOLS-2276

apupier commented:
> I anticipate that users will go to the source tab to edit their REST DSL XML. When they do so, that should update the underlying model. In the RestConfigurationEditor, line 158, I'm adding a model listener but it never gets triggered so far as I can tell. Even going back to CamelFile, line 429 (fireModelChanged) that doesn't seem to be fired when the source tab changes in debug mode.
> 
> What's the recommended pattern here for listening for changes on other pages of the editor and refreshing the Rest page?

the issue is that the RestConfigEditor.ctx attributes is not updated on reload (the mechanism to update the page when switching page was already provided)

apupier commented:
> We have a lot of flashing as it regenerates the items in the display - any suggestions on how to work around that?

I suggest to try to put a breakpoint in the reload method (or whatever is causing a redraw) and see what is calling it to check if some calls are not redundant.

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java:
```diff
@@ -532,17 +530,19 @@ private void refreshRestSection() {
 		toolkit.decorateFormHeading(form.getForm());
-		setSelection(StructuredSelection.EMPTY);
+		selection = null;
+		setSelection(new StructuredSelection());
```
why using new StructuredSelection() instead of StructuredSelection.EMPTY?
it was avoiding an object creation

apupier commented:
the test https://fusesource-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/jbosstools-fuse_PullRequest/1199/testReport/junit/org.fusesource.ide.camel.editor.integration/CamelEditorCheckOpeningInSpecialCasesIT/openContextFileAndCheckFor3Tabs/ needs to be updated

apupier commented:
when arriving to REST editor:
- the Rest element is selected but the properties view is not synchronized, there is a "Properties are not available
- when selecting one Rest Operations, going to design tab and then switching back to rest tab, the rest operation selection is lost. When there are several Rest Elements and that not the first one is selected, the selection is lost too.

apupier commented on editor/tests/org.fusesource.ide.wsdl2rest.ui.tests.integration/src/main/java/org/fusesource/ide/wsdl2rest/ui/tests/integration/Wsdl2RestWizardIT.java:
```diff
@@ -0,0 +1,162 @@
+ * 
+ * @author brianf
+ */
+public class Wsdl2RestWizardIT {
```
why this file is part of this PR? it seems strange

apupier commented on uitests/tests/org.jboss.tools.fuse.ui.bot.tests/resources/routeWithValidationFixedOnGlobalElements.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
```
why this file is part of this PR? it seems strange

bfitzpat replied:
Good question. I didn't change that file.

bfitzpat commented:
Created https://issues.jboss.org/browse/FUSETOOLS-2980 for selection issue

bfitzpat commented:
Seems the ID generation was saved silently under the covers. I added a check to see if the IDs changed -- if so, I mark the editor as dirty now. It's not the prettiest method, but it works. @apupier @lhein

bfitzpat commented:
testPR

bfitzpat commented:
After doing a few rounds of testing and playing with different options, you get the same behavior if you delete the id of a camel context element on the source tab, save it, tab off and tab back to the source tab. The ID gets generated I suspect when the model is saved and doesn't set the dirty flag. So the IDs problem is NOT an issue with the REST model code, this is more systemic and something we should probably talk about in a different JIRA. @lhein @apupier 

As a result, I have rolled back the change and left the code as it was before I modified it this morning. 

That just leaves the selection issue, which I will happily work on once this code gets merged.

apupier commented:
> As a result, I have rolled back the change

for next time, when "reverting/rolling back" a commit in a topic branch of a PR, I think it would be best to simply delete commit from history of the branch

apupier commented:
the failure on main PR job is due to an intermittent failing test which has been workarounded on master branch
the failures on Templates PR job is due to an issue with Fuse 7.0, which will be fixed with https://github.com/jbosstools/jbosstools-fuse/pull/1364

apupier commented on editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/properties/rest/RestVerbPropertiesFilter.java:
```diff
@@ -0,0 +1,50 @@
+	@Override
+	public boolean select(Object toTest) {
+		AbstractCamelModelElement ep = getSelectedObject(toTest);
+		List<String> restPropertyElements = new ArrayList<>();
```
optional: this list can be a static List, it will avoid to recreate a new one on each call

bfitzpat commented:
Huzzah!
